index.js

```diff
@@ -1,75 +1,154 @@
+  onerror = onerror || throwerr
+
+  const serialize = ndjson.serialize()
+  // TODO: Is there a race condition between the `time` config option and the request timeout option?
+  const chopper = new StreamChopper(opts).on('stream', onStream(opts, onerror))
+
+  // All sockets on the agent are unreffed when they are created. This means
+  // that when those are the only handles left, the `beforeExit` event will be
+  // emitted. By listening for this we can make sure to end the requests
+  // properly before exiting. This way we don't keep the process running until
+  // the `time` timeout happens.
+  // TODO: If you start too many clients (e.g. in a test), this will emit a too-many-listeners warning
+  process.once('beforeExit', function () {
+    opts.agent.ref() // re-ref the open socket handles
+    serialize.end()
+  })
 
-var SUB_USER_AGENT = pkg.name + '/' + pkg.version
+  pump(serialize, chopper, function (err) {
+    if (err) onerror(err)
+  })
 
-var Client = module.exports = function (opts) {
-  if (!(this instanceof Client)) return new Client(opts)
+  return serialize
+}
 
-  opts = opts || {}
+function onStream (opts, onerror) {
+  const transport = opts.transport
+  const serverTimeout = opts.serverTimeout
+  opts = getRequestOptions(opts)
 
-  if (!opts.userAgent) throw new Error('Missing required option: userAgent')
+  return function (stream, next) {
+    const req = transport.request(opts, onResult(onerror))
 
-  this.secretToken = opts.secretToken || null
-  this.userAgent = opts.userAgent + ' ' + SUB_USER_AGENT
+    req.on('socket', function (socket) {
+      // Sockets will automatically be unreffed by the HTTP agent when they are
+      // not in use by an HTTP request, but as we're keeping the HTTP request
+      // open, we need to unref the socket manually
+      socket.unref()
+    })
 
-  var url = parseUrl(opts.serverUrl || 'http://localhost:8200')
+    // TODO: It doesn't really make sense to set req.serverTimeout on a
+    // streaming request before all the data is sent. We need find a solution
+    // for how to handle timeout of the response.
+    if (Number.isFinite(serverTimeout)) {
+      req.setTimeout(serverTimeout, function () {
+        req.abort()
+      })
+    }
 
-  this._api = {
-    hostname: url.hostname,
-    port: url.port,
-    transport: url.protocol === 'https:' ? require('https') : http,
-    path: url.path === '/' ? '/v1/' : url.path + '/v1/',
-    rejectUnauthorized: opts.rejectUnauthorized !== false,
-    serverTimeout: opts.serverTimeout
+    pump(stream, zlib.createGzip(), req, function (err) {
+      if (err) onerror(err)
+      // Add listener for req errors in case a timeout occurs AFTER the request
+      // finishes
+      //
+      // TODO: This will produce two errors, which might not be that nice
+      // - First one is caught by the if-sentence above (premature closure)
+      // - Second one is caught by the listener below (socket hung up)
+      req.on('error', onerror)
```

Comment:
Can you explain why the error event handler is attached here rather than before the pump?

Reply:
The `pump` module will also catch the first error. So if we add it before the pump we'll get the first error twice. But I agree this isn't nice. But that's more down to the fact that the core http module is a bit of a mess at times. It can emit more than one error on the `req` object - even after the request have finished. The 2nd error it emits is actually not coming from the request, but it just uses that object to emit the error on 🤷‍♂️

Reply:
If you have a better idea I'm all 👂s

Reply:
Perhaps rather than running `onerror` from pump it could _only_ do that via a `req.on('error, onerror)` outside the pump call? Then the pump callback would _only_ call `next`. That should result in only one error path, right? Does pump forward the actual error events across the streams, or does it only aggregate the errors into the callback?

Reply:
Fixed